{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "local_EGAD_fastq_folder = \"../../data/public_data_sets/EGAD00001006441/RNAseq_data\"\n",
    "linstat_EGAD_fastq_folder = \"/project/rnaseq/EGAD00001006441\"\n",
    "rnaseq_samplesheet_path = \"../../data/public_data_sets/EGAD00001006441/rnaseq_samplesheet.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create RNA-seq sample sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>lane_number</th>\n",
       "      <th>read_number</th>\n",
       "      <th>relative_path</th>\n",
       "      <th>absolute_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CL1</td>\n",
       "      <td>L001</td>\n",
       "      <td>R1</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CL1</td>\n",
       "      <td>L001</td>\n",
       "      <td>R2</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CL1</td>\n",
       "      <td>L002</td>\n",
       "      <td>R1</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CL1</td>\n",
       "      <td>L002</td>\n",
       "      <td>R2</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CL1</td>\n",
       "      <td>L003</td>\n",
       "      <td>R1</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CL7</td>\n",
       "      <td>L002</td>\n",
       "      <td>R2</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>CL7</td>\n",
       "      <td>L003</td>\n",
       "      <td>R1</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CL7</td>\n",
       "      <td>L003</td>\n",
       "      <td>R2</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CL7</td>\n",
       "      <td>L004</td>\n",
       "      <td>R1</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CL7</td>\n",
       "      <td>L004</td>\n",
       "      <td>R2</td>\n",
       "      <td>../../data/public_data_sets/EGAD00001006441/RN...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name lane_number read_number  \\\n",
       "33         CL1        L001          R1   \n",
       "13         CL1        L001          R2   \n",
       "20         CL1        L002          R1   \n",
       "51         CL1        L002          R2   \n",
       "17         CL1        L003          R1   \n",
       "..         ...         ...         ...   \n",
       "38         CL7        L002          R2   \n",
       "58         CL7        L003          R1   \n",
       "37         CL7        L003          R2   \n",
       "8          CL7        L004          R1   \n",
       "9          CL7        L004          R2   \n",
       "\n",
       "                                        relative_path  \\\n",
       "33  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "13  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "20  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "51  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "17  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "..                                                ...   \n",
       "38  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "58  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "37  ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "8   ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "9   ../../data/public_data_sets/EGAD00001006441/RN...   \n",
       "\n",
       "                                        absolute_path  \n",
       "33  /project/rnaseq/EGAD00001006441/EGAF0000471980...  \n",
       "13  /project/rnaseq/EGAD00001006441/EGAF0000471980...  \n",
       "20  /project/rnaseq/EGAD00001006441/EGAF0000471980...  \n",
       "51  /project/rnaseq/EGAD00001006441/EGAF0000471980...  \n",
       "17  /project/rnaseq/EGAD00001006441/EGAF0000471980...  \n",
       "..                                                ...  \n",
       "38  /project/rnaseq/EGAD00001006441/EGAF0000471983...  \n",
       "58  /project/rnaseq/EGAD00001006441/EGAF0000471984...  \n",
       "37  /project/rnaseq/EGAD00001006441/EGAF0000471984...  \n",
       "8   /project/rnaseq/EGAD00001006441/EGAF0000471984...  \n",
       "9   /project/rnaseq/EGAD00001006441/EGAF0000471984...  \n",
       "\n",
       "[64 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fastq_file_info = pd.DataFrame(columns=[\"sample_name\", \"lane_number\", \"read_number\", \"relative_path\", \"absolute_path\"])\n",
    "\n",
    "# Loop over the folders inside EGAD_fastq_folder\n",
    "for sub_folder in os.listdir(local_EGAD_fastq_folder):\n",
    "    # Check if the folder is a directory\n",
    "    if os.path.isdir(os.path.join(local_EGAD_fastq_folder, sub_folder)):\n",
    "        # Loop over the files inside the folder\n",
    "        for file in os.listdir(os.path.join(local_EGAD_fastq_folder, sub_folder)):\n",
    "            # Check if the file is a fastq file\n",
    "            if file.endswith(\".fastq.gz\"):\n",
    "                rel_path = os.path.join(local_EGAD_fastq_folder, sub_folder, file)\n",
    "                abs_path = os.path.join(linstat_EGAD_fastq_folder, sub_folder, file)\n",
    "\n",
    "                # Parse sample info out of filename\n",
    "                # Format of filename: samplename-Carey_lanenumber_readnumber_001.fastq.gz\n",
    "                filename = file.split(\".\")[0]\n",
    "                fastq_file_info.loc[len(fastq_file_info)] = {\n",
    "                    \"sample_name\": filename.split(\"-\")[0],\n",
    "                    \"lane_number\": filename.split(\"_\")[2],\n",
    "                    \"read_number\": filename.split(\"_\")[3],\n",
    "                    \"relative_path\": rel_path,\n",
    "                    \"absolute_path\": abs_path\n",
    "                }\n",
    "fastq_file_info = fastq_file_info.sort_values(by=[\"sample_name\", \"lane_number\", \"read_number\"])\n",
    "display(fastq_file_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>fastq_1</th>\n",
       "      <th>fastq_2</th>\n",
       "      <th>strandedness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CL1</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL1</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CL1</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CL1</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CL10</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CL10</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CL10</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CL10</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CL14</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471978...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CL14</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CL14</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CL14</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CL15</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CL15</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471979...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CL15</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CL15</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471980...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CL2</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CL2</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CL2</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CL2</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471981...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CL3</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CL3</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CL3</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CL3</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CL6</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471982...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CL6</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CL6</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CL6</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CL7</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CL7</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471983...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CL7</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CL7</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "      <td>/project/rnaseq/EGAD00001006441/EGAF0000471984...</td>\n",
       "      <td>auto</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample                                            fastq_1  \\\n",
       "0     CL1  /project/rnaseq/EGAD00001006441/EGAF0000471980...   \n",
       "1     CL1  /project/rnaseq/EGAD00001006441/EGAF0000471980...   \n",
       "2     CL1  /project/rnaseq/EGAD00001006441/EGAF0000471980...   \n",
       "3     CL1  /project/rnaseq/EGAD00001006441/EGAF0000471981...   \n",
       "4    CL10  /project/rnaseq/EGAD00001006441/EGAF0000471978...   \n",
       "5    CL10  /project/rnaseq/EGAD00001006441/EGAF0000471978...   \n",
       "6    CL10  /project/rnaseq/EGAD00001006441/EGAF0000471978...   \n",
       "7    CL10  /project/rnaseq/EGAD00001006441/EGAF0000471978...   \n",
       "8    CL14  /project/rnaseq/EGAD00001006441/EGAF0000471978...   \n",
       "9    CL14  /project/rnaseq/EGAD00001006441/EGAF0000471979...   \n",
       "10   CL14  /project/rnaseq/EGAD00001006441/EGAF0000471979...   \n",
       "11   CL14  /project/rnaseq/EGAD00001006441/EGAF0000471979...   \n",
       "12   CL15  /project/rnaseq/EGAD00001006441/EGAF0000471979...   \n",
       "13   CL15  /project/rnaseq/EGAD00001006441/EGAF0000471979...   \n",
       "14   CL15  /project/rnaseq/EGAD00001006441/EGAF0000471980...   \n",
       "15   CL15  /project/rnaseq/EGAD00001006441/EGAF0000471980...   \n",
       "16    CL2  /project/rnaseq/EGAD00001006441/EGAF0000471981...   \n",
       "17    CL2  /project/rnaseq/EGAD00001006441/EGAF0000471981...   \n",
       "18    CL2  /project/rnaseq/EGAD00001006441/EGAF0000471981...   \n",
       "19    CL2  /project/rnaseq/EGAD00001006441/EGAF0000471981...   \n",
       "20    CL3  /project/rnaseq/EGAD00001006441/EGAF0000471982...   \n",
       "21    CL3  /project/rnaseq/EGAD00001006441/EGAF0000471982...   \n",
       "22    CL3  /project/rnaseq/EGAD00001006441/EGAF0000471982...   \n",
       "23    CL3  /project/rnaseq/EGAD00001006441/EGAF0000471982...   \n",
       "24    CL6  /project/rnaseq/EGAD00001006441/EGAF0000471982...   \n",
       "25    CL6  /project/rnaseq/EGAD00001006441/EGAF0000471983...   \n",
       "26    CL6  /project/rnaseq/EGAD00001006441/EGAF0000471983...   \n",
       "27    CL6  /project/rnaseq/EGAD00001006441/EGAF0000471983...   \n",
       "28    CL7  /project/rnaseq/EGAD00001006441/EGAF0000471983...   \n",
       "29    CL7  /project/rnaseq/EGAD00001006441/EGAF0000471983...   \n",
       "30    CL7  /project/rnaseq/EGAD00001006441/EGAF0000471984...   \n",
       "31    CL7  /project/rnaseq/EGAD00001006441/EGAF0000471984...   \n",
       "\n",
       "                                              fastq_2 strandedness  \n",
       "0   /project/rnaseq/EGAD00001006441/EGAF0000471980...         auto  \n",
       "1   /project/rnaseq/EGAD00001006441/EGAF0000471980...         auto  \n",
       "2   /project/rnaseq/EGAD00001006441/EGAF0000471980...         auto  \n",
       "3   /project/rnaseq/EGAD00001006441/EGAF0000471981...         auto  \n",
       "4   /project/rnaseq/EGAD00001006441/EGAF0000471978...         auto  \n",
       "5   /project/rnaseq/EGAD00001006441/EGAF0000471978...         auto  \n",
       "6   /project/rnaseq/EGAD00001006441/EGAF0000471978...         auto  \n",
       "7   /project/rnaseq/EGAD00001006441/EGAF0000471978...         auto  \n",
       "8   /project/rnaseq/EGAD00001006441/EGAF0000471978...         auto  \n",
       "9   /project/rnaseq/EGAD00001006441/EGAF0000471979...         auto  \n",
       "10  /project/rnaseq/EGAD00001006441/EGAF0000471979...         auto  \n",
       "11  /project/rnaseq/EGAD00001006441/EGAF0000471979...         auto  \n",
       "12  /project/rnaseq/EGAD00001006441/EGAF0000471979...         auto  \n",
       "13  /project/rnaseq/EGAD00001006441/EGAF0000471979...         auto  \n",
       "14  /project/rnaseq/EGAD00001006441/EGAF0000471980...         auto  \n",
       "15  /project/rnaseq/EGAD00001006441/EGAF0000471980...         auto  \n",
       "16  /project/rnaseq/EGAD00001006441/EGAF0000471981...         auto  \n",
       "17  /project/rnaseq/EGAD00001006441/EGAF0000471981...         auto  \n",
       "18  /project/rnaseq/EGAD00001006441/EGAF0000471981...         auto  \n",
       "19  /project/rnaseq/EGAD00001006441/EGAF0000471981...         auto  \n",
       "20  /project/rnaseq/EGAD00001006441/EGAF0000471982...         auto  \n",
       "21  /project/rnaseq/EGAD00001006441/EGAF0000471982...         auto  \n",
       "22  /project/rnaseq/EGAD00001006441/EGAF0000471982...         auto  \n",
       "23  /project/rnaseq/EGAD00001006441/EGAF0000471982...         auto  \n",
       "24  /project/rnaseq/EGAD00001006441/EGAF0000471982...         auto  \n",
       "25  /project/rnaseq/EGAD00001006441/EGAF0000471983...         auto  \n",
       "26  /project/rnaseq/EGAD00001006441/EGAF0000471983...         auto  \n",
       "27  /project/rnaseq/EGAD00001006441/EGAF0000471983...         auto  \n",
       "28  /project/rnaseq/EGAD00001006441/EGAF0000471983...         auto  \n",
       "29  /project/rnaseq/EGAD00001006441/EGAF0000471983...         auto  \n",
       "30  /project/rnaseq/EGAD00001006441/EGAF0000471984...         auto  \n",
       "31  /project/rnaseq/EGAD00001006441/EGAF0000471984...         auto  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Validate that each unique (sample_name, lane_number) pair has exactly 2 reads\n",
    "grouped = fastq_file_info.groupby(['sample_name', 'lane_number'])['read_number'].apply(set)\n",
    "invalid_groups = grouped[grouped.apply(lambda x: x != {\"R1\", \"R2\"})]\n",
    "assert len(invalid_groups) == 0, f\"The following (sample_name, lane_number) pairs do not have exactly one 'R1' and one 'R2': {invalid_groups}\"\n",
    "\n",
    "# Reshape the dataframe to have one row per (sample_lane_number) pair\n",
    "pivot_df = fastq_file_info.pivot(index=['sample_name', 'lane_number'], columns='read_number', values='absolute_path')\n",
    "pivot_df.reset_index(inplace=True)\n",
    "\n",
    "# Format rnaseq samplesheet\n",
    "rnaseq_samplesheet = pd.DataFrame({\n",
    "    \"sample\": pivot_df[\"sample_name\"],\n",
    "    \"fastq_1\": pivot_df[\"R1\"],\n",
    "    \"fastq_2\": pivot_df[\"R2\"],\n",
    "    \"strandedness\": \"auto\"\n",
    "})\n",
    "\n",
    "display(rnaseq_samplesheet)\n",
    "\n",
    "# Write rnaseq samplesheet to file\n",
    "rnaseq_samplesheet.to_csv(rnaseq_samplesheet_path, index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run rnaseq pipeline\n",
    "\n",
    "```{sh}\n",
    "./nextflow run nf-core/rnaseq \\\n",
    "    --input /project9/rnaseq/EGAD00001006441/rnaseq_samplesheet.csv \\\n",
    "    --outdir EGAD00001006441 \\\n",
    "    -N -bg \\\n",
    "    --igenomes_base /project/rnaseq/ngi-igenomes/igenomes \\\n",
    "    --genome GRCh38\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
