{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import biomart      \n",
    "\n",
    "# Import utilities\n",
    "import os\n",
    "import sys\n",
    "os.chdir(\"/Volumes/kueck/PublicDataAnalysis/CASCAM_style_subtype_classification/src/preprocessing\")\n",
    "# os.getcwd() not working as expected, so need to set wd manually (update as needed).\n",
    "parent_dir = os.path.abspath(os.path.join(os.getcwd(), '../'))\n",
    "sys.path.append(parent_dir)\n",
    "\n",
    "from utilities import get_ensembl_mappings_grch38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipped 22300 entries because they were missing gene_symbol\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48306</th>\n",
       "      <td>ENSG00000232679</td>\n",
       "      <td>LINC01705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48307</th>\n",
       "      <td>ENSG00000200033</td>\n",
       "      <td>RNU6-403P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48308</th>\n",
       "      <td>ENSG00000228437</td>\n",
       "      <td>LINC02474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48309</th>\n",
       "      <td>ENSG00000229463</td>\n",
       "      <td>LYST-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48310</th>\n",
       "      <td>ENSG00000229291</td>\n",
       "      <td>LINC02768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48311 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name\n",
       "0      ENSG00000210049              MT-TF\n",
       "1      ENSG00000211459            MT-RNR1\n",
       "2      ENSG00000210077              MT-TV\n",
       "3      ENSG00000210082            MT-RNR2\n",
       "4      ENSG00000209082             MT-TL1\n",
       "...                ...                ...\n",
       "48306  ENSG00000232679          LINC01705\n",
       "48307  ENSG00000200033          RNU6-403P\n",
       "48308  ENSG00000228437          LINC02474\n",
       "48309  ENSG00000229463           LYST-AS1\n",
       "48310  ENSG00000229291          LINC02768\n",
       "\n",
       "[48311 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ensembl_to_gene_symbol_df = get_ensembl_mappings_grch38()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcga_data_folder = \"../../data/public_data_sets/TCGA_OV_RNA\"\n",
    "log2_counts_file = f\"{tcga_data_folder}/TCGA-OV.htseq_counts.tsv\"\n",
    "mutations_file = f\"{tcga_data_folder}/TCGA-OV.muse_snv.tsv\"\n",
    "copy_number_file = f\"{tcga_data_folder}/TCGA-OV.gistic.tsv\"\n",
    "metadata_file = f\"{tcga_data_folder}/TCGA_cohort_metadata.json\"\n",
    "parsed_metadata_file = f\"{tcga_data_folder}/TCGA_cohort_metadata_parsed.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and parse metadata file\n",
    "\n",
    "Creates txt file from json metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primary_site</th>\n",
       "      <th>submitter_slide_ids</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>case_id</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>summary.file_count</th>\n",
       "      <th>summary.experimental_strategies</th>\n",
       "      <th>project.project_id</th>\n",
       "      <th>project.program.name</th>\n",
       "      <th>demographic.race</th>\n",
       "      <th>demographic.gender</th>\n",
       "      <th>demographic.ethnicity</th>\n",
       "      <th>demographic.vital_status</th>\n",
       "      <th>demographic.days_to_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-0920-01A-01-BS1, TCGA-13-0920-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>85a85a11-7200-4e96-97af-6ba26d680d59</td>\n",
       "      <td>TCGA-13-0920</td>\n",
       "      <td>[{'age_at_diagnosis': 24064}]</td>\n",
       "      <td>84</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1484.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-09-1662-01A-01-BS1, TCGA-09-1662-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>867f9563-16c9-45a8-b519-6df61ba1b6b7</td>\n",
       "      <td>TCGA-09-1662</td>\n",
       "      <td>[{'age_at_diagnosis': 21246}]</td>\n",
       "      <td>54</td>\n",
       "      <td>[miRNA-Seq, RNA-Seq, WGS, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>2717.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-24-2033-01A-01-BS1, TCGA-24-2033-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>872d2922-7292-4681-adb7-d3b267eccbe7</td>\n",
       "      <td>TCGA-24-2033</td>\n",
       "      <td>[{'age_at_diagnosis': 31977}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>562.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-61-2110-11A-01-TS1, TCGA-61-2110-11A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>88180134-710f-46ea-9b06-5e5d860d6d9f</td>\n",
       "      <td>TCGA-61-2110</td>\n",
       "      <td>[{'age_at_diagnosis': 20718}]</td>\n",
       "      <td>69</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-29-1696-01A-01-TS1, TCGA-29-1696-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>8a98a6e6-b763-4824-858b-fd2738e6c9a3</td>\n",
       "      <td>TCGA-29-1696</td>\n",
       "      <td>[{'age_at_diagnosis': 15818}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1032.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-10-0938-01A-01-BS1, TCGA-10-0938-01A-02-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>7ca97692-0bf5-4bbd-81ce-10a051d04bd5</td>\n",
       "      <td>TCGA-10-0938</td>\n",
       "      <td>[{'age_at_diagnosis': 29558}]</td>\n",
       "      <td>53</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>636.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-0804-01A-01-TS1, TCGA-13-0804-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>7e34d3c1-1fab-4326-9a69-4260d2bac558</td>\n",
       "      <td>TCGA-13-0804</td>\n",
       "      <td>[{'age_at_diagnosis': 27021}]</td>\n",
       "      <td>84</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1073.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-25-2042-01A-01-BS1, TCGA-25-2042-01Z-00-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>818dc159-aba4-46bc-a4ed-68ee0f8c4461</td>\n",
       "      <td>TCGA-25-2042</td>\n",
       "      <td>[{'age_at_diagnosis': 22127}]</td>\n",
       "      <td>69</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>american indian or alaska native</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>396.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-1512-01A-01-TS1, TCGA-13-1512-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>82093ed9-a3c8-4e34-931f-4ec7ae745711</td>\n",
       "      <td>TCGA-13-1512</td>\n",
       "      <td>[{'age_at_diagnosis': 18134}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-29-1768-01A-01-BS1, TCGA-29-1768-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>838693d5-1ae0-4d75-834b-e1b89b96b0ed</td>\n",
       "      <td>TCGA-29-1768</td>\n",
       "      <td>[{'age_at_diagnosis': 18396}]</td>\n",
       "      <td>68</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>952.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    primary_site                                submitter_slide_ids  \\\n",
       "0          Ovary  [TCGA-13-0920-01A-01-BS1, TCGA-13-0920-01A-01-...   \n",
       "1          Ovary  [TCGA-09-1662-01A-01-BS1, TCGA-09-1662-01A-01-...   \n",
       "2          Ovary  [TCGA-24-2033-01A-01-BS1, TCGA-24-2033-01A-01-...   \n",
       "3          Ovary  [TCGA-61-2110-11A-01-TS1, TCGA-61-2110-11A-01-...   \n",
       "4          Ovary  [TCGA-29-1696-01A-01-TS1, TCGA-29-1696-01A-01-...   \n",
       "..           ...                                                ...   \n",
       "417        Ovary  [TCGA-10-0938-01A-01-BS1, TCGA-10-0938-01A-02-...   \n",
       "418        Ovary  [TCGA-13-0804-01A-01-TS1, TCGA-13-0804-01A-01-...   \n",
       "419        Ovary  [TCGA-25-2042-01A-01-BS1, TCGA-25-2042-01Z-00-...   \n",
       "420        Ovary  [TCGA-13-1512-01A-01-TS1, TCGA-13-1512-01A-01-...   \n",
       "421        Ovary  [TCGA-29-1768-01A-01-BS1, TCGA-29-1768-01A-01-...   \n",
       "\n",
       "                              disease_type  \\\n",
       "0    Cystic, Mucinous and Serous Neoplasms   \n",
       "1    Cystic, Mucinous and Serous Neoplasms   \n",
       "2    Cystic, Mucinous and Serous Neoplasms   \n",
       "3    Cystic, Mucinous and Serous Neoplasms   \n",
       "4    Cystic, Mucinous and Serous Neoplasms   \n",
       "..                                     ...   \n",
       "417  Cystic, Mucinous and Serous Neoplasms   \n",
       "418  Cystic, Mucinous and Serous Neoplasms   \n",
       "419  Cystic, Mucinous and Serous Neoplasms   \n",
       "420  Cystic, Mucinous and Serous Neoplasms   \n",
       "421  Cystic, Mucinous and Serous Neoplasms   \n",
       "\n",
       "                                  case_id  submitter_id  \\\n",
       "0    85a85a11-7200-4e96-97af-6ba26d680d59  TCGA-13-0920   \n",
       "1    867f9563-16c9-45a8-b519-6df61ba1b6b7  TCGA-09-1662   \n",
       "2    872d2922-7292-4681-adb7-d3b267eccbe7  TCGA-24-2033   \n",
       "3    88180134-710f-46ea-9b06-5e5d860d6d9f  TCGA-61-2110   \n",
       "4    8a98a6e6-b763-4824-858b-fd2738e6c9a3  TCGA-29-1696   \n",
       "..                                    ...           ...   \n",
       "417  7ca97692-0bf5-4bbd-81ce-10a051d04bd5  TCGA-10-0938   \n",
       "418  7e34d3c1-1fab-4326-9a69-4260d2bac558  TCGA-13-0804   \n",
       "419  818dc159-aba4-46bc-a4ed-68ee0f8c4461  TCGA-25-2042   \n",
       "420  82093ed9-a3c8-4e34-931f-4ec7ae745711  TCGA-13-1512   \n",
       "421  838693d5-1ae0-4d75-834b-e1b89b96b0ed  TCGA-29-1768   \n",
       "\n",
       "                         diagnoses  summary.file_count  \\\n",
       "0    [{'age_at_diagnosis': 24064}]                  84   \n",
       "1    [{'age_at_diagnosis': 21246}]                  54   \n",
       "2    [{'age_at_diagnosis': 31977}]                  70   \n",
       "3    [{'age_at_diagnosis': 20718}]                  69   \n",
       "4    [{'age_at_diagnosis': 15818}]                  70   \n",
       "..                             ...                 ...   \n",
       "417  [{'age_at_diagnosis': 29558}]                  53   \n",
       "418  [{'age_at_diagnosis': 27021}]                  84   \n",
       "419  [{'age_at_diagnosis': 22127}]                  69   \n",
       "420  [{'age_at_diagnosis': 18134}]                  70   \n",
       "421  [{'age_at_diagnosis': 18396}]                  68   \n",
       "\n",
       "                       summary.experimental_strategies project.project_id  \\\n",
       "0    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "1    [miRNA-Seq, RNA-Seq, WGS, Genotyping Array, Me...            TCGA-OV   \n",
       "2    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "3    [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "4    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "..                                                 ...                ...   \n",
       "417  [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "418  [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "419  [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "420  [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "421  [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "\n",
       "    project.program.name                  demographic.race demographic.gender  \\\n",
       "0                   TCGA                             white             female   \n",
       "1                   TCGA                             white             female   \n",
       "2                   TCGA                             white             female   \n",
       "3                   TCGA                             white             female   \n",
       "4                   TCGA                             white             female   \n",
       "..                   ...                               ...                ...   \n",
       "417                 TCGA                             white             female   \n",
       "418                 TCGA                             white             female   \n",
       "419                 TCGA  american indian or alaska native             female   \n",
       "420                 TCGA                             white             female   \n",
       "421                 TCGA                             white             female   \n",
       "\n",
       "      demographic.ethnicity demographic.vital_status  \\\n",
       "0    not hispanic or latino                     Dead   \n",
       "1    not hispanic or latino                     Dead   \n",
       "2              not reported                     Dead   \n",
       "3    not hispanic or latino                     Dead   \n",
       "4              not reported                     Dead   \n",
       "..                      ...                      ...   \n",
       "417            not reported                     Dead   \n",
       "418  not hispanic or latino                     Dead   \n",
       "419  not hispanic or latino                     Dead   \n",
       "420  not hispanic or latino                    Alive   \n",
       "421            not reported                     Dead   \n",
       "\n",
       "     demographic.days_to_death  \n",
       "0                       1484.0  \n",
       "1                       2717.0  \n",
       "2                        562.0  \n",
       "3                       1354.0  \n",
       "4                       1032.0  \n",
       "..                         ...  \n",
       "417                      636.0  \n",
       "418                     1073.0  \n",
       "419                      396.0  \n",
       "420                        NaN  \n",
       "421                      952.0  \n",
       "\n",
       "[422 rows x 15 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with open(metadata_file, \"r\") as file:\n",
    "    metadata_json = json.loads(file.read())\n",
    "metadata = pd.json_normalize(metadata_json)\n",
    "\n",
    "# Format the experimental strategies more legibly by parsing list of dictionaries\n",
    "metadata[\"summary.experimental_strategies\"] = metadata[\n",
    "    \"summary.experimental_strategies\"\n",
    "].apply(lambda exp_list: [exp[\"experimental_strategy\"] for exp in exp_list])\n",
    "\n",
    "display(metadata)\n",
    "\n",
    "metadata.to_csv(parsed_metadata_file, sep = \"\\t\", index=False)\n",
    "\n",
    "# Check that all submitter ids are unique\n",
    "assert metadata[\"submitter_id\"].nunique() == metadata.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load log2 counts data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>TCGA-04-1331</th>\n",
       "      <th>TCGA-04-1332</th>\n",
       "      <th>TCGA-04-1338</th>\n",
       "      <th>TCGA-04-1341</th>\n",
       "      <th>TCGA-04-1343</th>\n",
       "      <th>TCGA-04-1347</th>\n",
       "      <th>TCGA-04-1350</th>\n",
       "      <th>TCGA-04-1356</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-61-2101</th>\n",
       "      <th>TCGA-61-2102</th>\n",
       "      <th>TCGA-61-2104</th>\n",
       "      <th>TCGA-61-2109</th>\n",
       "      <th>TCGA-61-2110</th>\n",
       "      <th>TCGA-61-2111</th>\n",
       "      <th>TCGA-61-2113</th>\n",
       "      <th>TCGA-OY-A56Q</th>\n",
       "      <th>TCGA-VG-A8LO</th>\n",
       "      <th>TCGA-WR-A838</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>12.804131</td>\n",
       "      <td>11.340963</td>\n",
       "      <td>10.324181</td>\n",
       "      <td>11.982637</td>\n",
       "      <td>11.885696</td>\n",
       "      <td>12.466586</td>\n",
       "      <td>12.219169</td>\n",
       "      <td>12.427051</td>\n",
       "      <td>...</td>\n",
       "      <td>11.991876</td>\n",
       "      <td>11.413099</td>\n",
       "      <td>11.516685</td>\n",
       "      <td>11.423641</td>\n",
       "      <td>9.952741</td>\n",
       "      <td>11.976922</td>\n",
       "      <td>11.887982</td>\n",
       "      <td>11.969746</td>\n",
       "      <td>12.122828</td>\n",
       "      <td>12.583318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>7.546894</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>3.169925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>11.669771</td>\n",
       "      <td>10.344296</td>\n",
       "      <td>10.451211</td>\n",
       "      <td>11.791163</td>\n",
       "      <td>13.057146</td>\n",
       "      <td>12.323899</td>\n",
       "      <td>11.459944</td>\n",
       "      <td>11.805341</td>\n",
       "      <td>...</td>\n",
       "      <td>12.127350</td>\n",
       "      <td>10.632995</td>\n",
       "      <td>11.373409</td>\n",
       "      <td>10.764042</td>\n",
       "      <td>10.800091</td>\n",
       "      <td>11.249706</td>\n",
       "      <td>11.759472</td>\n",
       "      <td>11.455841</td>\n",
       "      <td>11.658211</td>\n",
       "      <td>11.528942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.12</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>10.364135</td>\n",
       "      <td>8.797662</td>\n",
       "      <td>8.566054</td>\n",
       "      <td>9.301496</td>\n",
       "      <td>9.036174</td>\n",
       "      <td>9.071462</td>\n",
       "      <td>8.693487</td>\n",
       "      <td>9.741467</td>\n",
       "      <td>...</td>\n",
       "      <td>10.273796</td>\n",
       "      <td>9.957102</td>\n",
       "      <td>10.381543</td>\n",
       "      <td>9.703904</td>\n",
       "      <td>9.398744</td>\n",
       "      <td>10.135709</td>\n",
       "      <td>9.968667</td>\n",
       "      <td>10.736402</td>\n",
       "      <td>10.729621</td>\n",
       "      <td>10.222795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.15</th>\n",
       "      <td>FIRRM</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>9.812177</td>\n",
       "      <td>8.471675</td>\n",
       "      <td>7.569856</td>\n",
       "      <td>8.632995</td>\n",
       "      <td>8.299208</td>\n",
       "      <td>8.092757</td>\n",
       "      <td>7.988685</td>\n",
       "      <td>8.985842</td>\n",
       "      <td>...</td>\n",
       "      <td>9.794416</td>\n",
       "      <td>10.203348</td>\n",
       "      <td>9.541097</td>\n",
       "      <td>9.503826</td>\n",
       "      <td>8.876517</td>\n",
       "      <td>9.245553</td>\n",
       "      <td>9.954196</td>\n",
       "      <td>9.477758</td>\n",
       "      <td>9.917372</td>\n",
       "      <td>9.533330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>__no_feature</th>\n",
       "      <td>None</td>\n",
       "      <td>__no_feature</td>\n",
       "      <td>23.565784</td>\n",
       "      <td>23.194738</td>\n",
       "      <td>22.149820</td>\n",
       "      <td>22.474268</td>\n",
       "      <td>22.828578</td>\n",
       "      <td>22.161203</td>\n",
       "      <td>22.279944</td>\n",
       "      <td>23.237803</td>\n",
       "      <td>...</td>\n",
       "      <td>22.598964</td>\n",
       "      <td>21.584929</td>\n",
       "      <td>22.804893</td>\n",
       "      <td>22.821447</td>\n",
       "      <td>22.747116</td>\n",
       "      <td>22.056261</td>\n",
       "      <td>23.035873</td>\n",
       "      <td>23.152011</td>\n",
       "      <td>23.547087</td>\n",
       "      <td>23.248577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>__ambiguous</th>\n",
       "      <td>None</td>\n",
       "      <td>__ambiguous</td>\n",
       "      <td>21.348947</td>\n",
       "      <td>20.827971</td>\n",
       "      <td>20.330253</td>\n",
       "      <td>22.516980</td>\n",
       "      <td>22.132119</td>\n",
       "      <td>21.377337</td>\n",
       "      <td>21.835249</td>\n",
       "      <td>21.728927</td>\n",
       "      <td>...</td>\n",
       "      <td>22.101431</td>\n",
       "      <td>20.672535</td>\n",
       "      <td>21.529985</td>\n",
       "      <td>21.100490</td>\n",
       "      <td>21.386072</td>\n",
       "      <td>20.646611</td>\n",
       "      <td>21.513571</td>\n",
       "      <td>21.506699</td>\n",
       "      <td>21.747044</td>\n",
       "      <td>21.519990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>__too_low_aQual</th>\n",
       "      <td>None</td>\n",
       "      <td>__too_low_aQual</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>__not_aligned</th>\n",
       "      <td>None</td>\n",
       "      <td>__not_aligned</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>__alignment_not_unique</th>\n",
       "      <td>None</td>\n",
       "      <td>__alignment_not_unique</td>\n",
       "      <td>25.853210</td>\n",
       "      <td>27.282577</td>\n",
       "      <td>25.428224</td>\n",
       "      <td>25.587177</td>\n",
       "      <td>25.531242</td>\n",
       "      <td>25.143135</td>\n",
       "      <td>26.035064</td>\n",
       "      <td>25.631160</td>\n",
       "      <td>...</td>\n",
       "      <td>25.947702</td>\n",
       "      <td>24.452542</td>\n",
       "      <td>25.420018</td>\n",
       "      <td>25.090680</td>\n",
       "      <td>25.584612</td>\n",
       "      <td>24.721992</td>\n",
       "      <td>25.037589</td>\n",
       "      <td>25.383720</td>\n",
       "      <td>24.881644</td>\n",
       "      <td>24.673386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60488 rows Ã— 378 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gene_symbol              ensembl_id  TCGA-04-1331  \\\n",
       "Ensembl_ID                                                                 \n",
       "ENSG00000000003.13          TSPAN6         ENSG00000000003     12.804131   \n",
       "ENSG00000000005.5             TNMD         ENSG00000000005      3.459432   \n",
       "ENSG00000000419.11            DPM1         ENSG00000000419     11.669771   \n",
       "ENSG00000000457.12           SCYL3         ENSG00000000457     10.364135   \n",
       "ENSG00000000460.15           FIRRM         ENSG00000000460      9.812177   \n",
       "...                            ...                     ...           ...   \n",
       "__no_feature                  None            __no_feature     23.565784   \n",
       "__ambiguous                   None             __ambiguous     21.348947   \n",
       "__too_low_aQual               None         __too_low_aQual      0.000000   \n",
       "__not_aligned                 None           __not_aligned      0.000000   \n",
       "__alignment_not_unique        None  __alignment_not_unique     25.853210   \n",
       "\n",
       "                        TCGA-04-1332  TCGA-04-1338  TCGA-04-1341  \\\n",
       "Ensembl_ID                                                         \n",
       "ENSG00000000003.13         11.340963     10.324181     11.982637   \n",
       "ENSG00000000005.5           3.700440      1.000000      3.169925   \n",
       "ENSG00000000419.11         10.344296     10.451211     11.791163   \n",
       "ENSG00000000457.12          8.797662      8.566054      9.301496   \n",
       "ENSG00000000460.15          8.471675      7.569856      8.632995   \n",
       "...                              ...           ...           ...   \n",
       "__no_feature               23.194738     22.149820     22.474268   \n",
       "__ambiguous                20.827971     20.330253     22.516980   \n",
       "__too_low_aQual             0.000000      0.000000      0.000000   \n",
       "__not_aligned               0.000000      0.000000      0.000000   \n",
       "__alignment_not_unique     27.282577     25.428224     25.587177   \n",
       "\n",
       "                        TCGA-04-1343  TCGA-04-1347  TCGA-04-1350  \\\n",
       "Ensembl_ID                                                         \n",
       "ENSG00000000003.13         11.885696     12.466586     12.219169   \n",
       "ENSG00000000005.5           7.546894      3.807355      0.000000   \n",
       "ENSG00000000419.11         13.057146     12.323899     11.459944   \n",
       "ENSG00000000457.12          9.036174      9.071462      8.693487   \n",
       "ENSG00000000460.15          8.299208      8.092757      7.988685   \n",
       "...                              ...           ...           ...   \n",
       "__no_feature               22.828578     22.161203     22.279944   \n",
       "__ambiguous                22.132119     21.377337     21.835249   \n",
       "__too_low_aQual             0.000000      0.000000      0.000000   \n",
       "__not_aligned               0.000000      0.000000      0.000000   \n",
       "__alignment_not_unique     25.531242     25.143135     26.035064   \n",
       "\n",
       "                        TCGA-04-1356  ...  TCGA-61-2101  TCGA-61-2102  \\\n",
       "Ensembl_ID                            ...                               \n",
       "ENSG00000000003.13         12.427051  ...     11.991876     11.413099   \n",
       "ENSG00000000005.5           1.584963  ...      1.584963      5.426265   \n",
       "ENSG00000000419.11         11.805341  ...     12.127350     10.632995   \n",
       "ENSG00000000457.12          9.741467  ...     10.273796      9.957102   \n",
       "ENSG00000000460.15          8.985842  ...      9.794416     10.203348   \n",
       "...                              ...  ...           ...           ...   \n",
       "__no_feature               23.237803  ...     22.598964     21.584929   \n",
       "__ambiguous                21.728927  ...     22.101431     20.672535   \n",
       "__too_low_aQual             0.000000  ...      0.000000      0.000000   \n",
       "__not_aligned               0.000000  ...      0.000000      0.000000   \n",
       "__alignment_not_unique     25.631160  ...     25.947702     24.452542   \n",
       "\n",
       "                        TCGA-61-2104  TCGA-61-2109  TCGA-61-2110  \\\n",
       "Ensembl_ID                                                         \n",
       "ENSG00000000003.13         11.516685     11.423641      9.952741   \n",
       "ENSG00000000005.5           1.000000      3.000000      2.584963   \n",
       "ENSG00000000419.11         11.373409     10.764042     10.800091   \n",
       "ENSG00000000457.12         10.381543      9.703904      9.398744   \n",
       "ENSG00000000460.15          9.541097      9.503826      8.876517   \n",
       "...                              ...           ...           ...   \n",
       "__no_feature               22.804893     22.821447     22.747116   \n",
       "__ambiguous                21.529985     21.100490     21.386072   \n",
       "__too_low_aQual             0.000000      0.000000      0.000000   \n",
       "__not_aligned               0.000000      0.000000      0.000000   \n",
       "__alignment_not_unique     25.420018     25.090680     25.584612   \n",
       "\n",
       "                        TCGA-61-2111  TCGA-61-2113  TCGA-OY-A56Q  \\\n",
       "Ensembl_ID                                                         \n",
       "ENSG00000000003.13         11.976922     11.887982     11.969746   \n",
       "ENSG00000000005.5           0.000000      1.584963      1.584963   \n",
       "ENSG00000000419.11         11.249706     11.759472     11.455841   \n",
       "ENSG00000000457.12         10.135709      9.968667     10.736402   \n",
       "ENSG00000000460.15          9.245553      9.954196      9.477758   \n",
       "...                              ...           ...           ...   \n",
       "__no_feature               22.056261     23.035873     23.152011   \n",
       "__ambiguous                20.646611     21.513571     21.506699   \n",
       "__too_low_aQual             0.000000      0.000000      0.000000   \n",
       "__not_aligned               0.000000      0.000000      0.000000   \n",
       "__alignment_not_unique     24.721992     25.037589     25.383720   \n",
       "\n",
       "                        TCGA-VG-A8LO  TCGA-WR-A838  \n",
       "Ensembl_ID                                          \n",
       "ENSG00000000003.13         12.122828     12.583318  \n",
       "ENSG00000000005.5           2.321928      3.169925  \n",
       "ENSG00000000419.11         11.658211     11.528942  \n",
       "ENSG00000000457.12         10.729621     10.222795  \n",
       "ENSG00000000460.15          9.917372      9.533330  \n",
       "...                              ...           ...  \n",
       "__no_feature               23.547087     23.248577  \n",
       "__ambiguous                21.747044     21.519990  \n",
       "__too_low_aQual             0.000000      0.000000  \n",
       "__not_aligned               0.000000      0.000000  \n",
       "__alignment_not_unique     24.881644     24.673386  \n",
       "\n",
       "[60488 rows x 378 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "log2_counts = pd.read_csv(log2_counts_file, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Rename with submitter ID instead of submitter_slide_id (up to 3rd \"-\"), merging samples with same submitter ID\n",
    "log2_counts.columns = ['-'.join(col.split(\"-\", 3)[:3]) for col in log2_counts.columns]\n",
    "log2_counts = log2_counts.T\n",
    "log2_counts = log2_counts.groupby(level=0).mean()\n",
    "log2_counts = log2_counts.T\n",
    "assert not log2_counts.columns.duplicated().any()\n",
    "\n",
    "# Add gene symbol column\n",
    "log2_counts.insert(0,\n",
    "                   \"ensembl_id\",\n",
    "                   [ensembl_id_with_version.split(\".\")[0] for ensembl_id_with_version in log2_counts.index])\n",
    "def get_gene_symbol(ensembl_id):\n",
    "    try:\n",
    "        return ensembl_to_gene_symbol_df.loc[ensembl_to_gene_symbol_df[\"ensembl_gene_id\"] == ensembl_id, \"external_gene_name\"].values[0]\n",
    "    except IndexError:\n",
    "        return None\n",
    "log2_counts.insert(0,\n",
    "                   \"gene_symbol\",\n",
    "                   log2_counts[\"ensembl_id\"].apply(get_gene_symbol))\n",
    "\n",
    "display(log2_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and parse mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num samples with TP53 mutations: 298/422\n",
      "Num samples with KRAS mutations: 5/422\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>tp53_mutant</th>\n",
       "      <th>kras_mutant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-13-0717</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-20-0996</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-13-1409</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-29-1766</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-24-2261</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>TCGA-13-0903</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>TCGA-25-2398</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>TCGA-31-1959</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>TCGA-31-1950</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>TCGA-24-2030</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>433 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     submitter_id  tp53_mutant  kras_mutant\n",
       "0    TCGA-13-0717        False        False\n",
       "1    TCGA-20-0996        False        False\n",
       "2    TCGA-13-1409         True        False\n",
       "3    TCGA-29-1766        False        False\n",
       "4    TCGA-24-2261        False        False\n",
       "..            ...          ...          ...\n",
       "428  TCGA-13-0903         True        False\n",
       "429  TCGA-25-2398        False        False\n",
       "430  TCGA-31-1959         True        False\n",
       "431  TCGA-31-1950         True        False\n",
       "432  TCGA-24-2030         True        False\n",
       "\n",
       "[433 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "individual_mutations = pd.read_csv(mutations_file, sep=\"\\t\")\n",
    "\n",
    "# Rename with submitter ID instead of sample_ID(up to 3rd \"-\")\n",
    "individual_mutations.insert(0, \"submitter_id\", ['-'.join(col.split(\"-\", 3)[:3]) for col in individual_mutations[\"Sample_ID\"]])\n",
    "\n",
    "# See what effects are present\n",
    "effects_nested = list(set(individual_mutations[\"effect\"]))\n",
    "effects = list(set([item for sublist in effects_nested for item in sublist.split(\";\")]))\n",
    "# display(pd.DataFrame({\"effect\": effects}))\n",
    "\n",
    "# Point mutation categories defined by Jessi\n",
    "point_mutant_effects = [\"stop_gained\",\n",
    "                       \"start_lost\",\n",
    "                       \"missense_variant\",\n",
    "                       \"splice_donor_variant\",\n",
    "                       \"splice_acceptor_variant\"]\n",
    "assert(all([effect in effects for effect in point_mutant_effects]))\n",
    "\n",
    "# Find TP53 mutants\n",
    "individual_mutations.insert(1, \"tp53_mutant\", individual_mutations[\"effect\"].apply(lambda effect: any([effect in point_mutant_effects])))\n",
    "tp53_point_mutants = list(set(individual_mutations[(individual_mutations[\"gene\"] == \"TP53\") & individual_mutations[\"tp53_mutant\"]][\"submitter_id\"]))\n",
    "\n",
    "# Identify KRAS mutants\n",
    "individual_mutations.insert(1, \"kras_mutant\", individual_mutations[\"effect\"].apply(lambda effect: any([effect in point_mutant_effects])))\n",
    "kras_point_mutants = list(set(individual_mutations[(individual_mutations[\"gene\"] == \"KRAS\") & individual_mutations[\"kras_mutant\"]][\"submitter_id\"]))\n",
    "\n",
    "print(f\"Num samples with TP53 mutations: {len(tp53_point_mutants)}/{metadata.shape[0]}\")\n",
    "print(f\"Num samples with KRAS mutations: {len(kras_point_mutants)}/{metadata.shape[0]}\")\n",
    "\n",
    "# Format into a data frame with submitter_id, tp53_mutant, kras_mutant\n",
    "mutations = pd.DataFrame(\n",
    "    {\n",
    "        \"submitter_id\": list(set(individual_mutations[\"submitter_id\"]))\n",
    "    }\n",
    ")\n",
    "mutations[\"tp53_mutant\"] = mutations[\"submitter_id\"].apply(lambda submitter_id: submitter_id in tp53_point_mutants)\n",
    "mutations[\"kras_mutant\"] = mutations[\"submitter_id\"].apply(lambda submitter_id: submitter_id in kras_point_mutants)\n",
    "\n",
    "# Remove redundant rows and ensure submitter ids are unique\n",
    "mutations = mutations.drop_duplicates(subset=[\"submitter_id\", \"tp53_mutant\", \"kras_mutant\"])\n",
    "assert len(set(mutations[\"submitter_id\"])) == len(mutations[\"submitter_id\"])\n",
    "display(mutations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and parse copy number data\n",
    "\n",
    "Convert to gene symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_id_with_version</th>\n",
       "      <th>TCGA-04-1517-01A</th>\n",
       "      <th>TCGA-24-1556-01A</th>\n",
       "      <th>TCGA-36-1571-01A</th>\n",
       "      <th>TCGA-57-1582-01A</th>\n",
       "      <th>TCGA-25-1628-01A</th>\n",
       "      <th>TCGA-09-1661-01B</th>\n",
       "      <th>TCGA-04-1519-01A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-72-4240-01A.1</th>\n",
       "      <th>TCGA-72-4232-01A.1</th>\n",
       "      <th>TCGA-72-4241-01A.1</th>\n",
       "      <th>TCGA-72-4233-01A.1</th>\n",
       "      <th>TCGA-72-4234-01A.1</th>\n",
       "      <th>TCGA-72-4235-01A.1</th>\n",
       "      <th>TCGA-72-4236-01A.1</th>\n",
       "      <th>TCGA-72-4237-01A.1</th>\n",
       "      <th>TCGA-72-4238-01A.1</th>\n",
       "      <th>has_gene_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDK11A</td>\n",
       "      <td>ENSG00000008128</td>\n",
       "      <td>ENSG00000008128.21</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NADK</td>\n",
       "      <td>ENSG00000008130</td>\n",
       "      <td>ENSG00000008130.14</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>ENSG00000067606.14</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GNB1</td>\n",
       "      <td>ENSG00000078369</td>\n",
       "      <td>ENSG00000078369.16</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SDF4</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>ENSG00000078808.15</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19724</th>\n",
       "      <td>H2AB3</td>\n",
       "      <td>ENSG00000277745</td>\n",
       "      <td>ENSG00000277745.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19725</th>\n",
       "      <td>H2AB2</td>\n",
       "      <td>ENSG00000277858</td>\n",
       "      <td>ENSG00000277858.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19726</th>\n",
       "      <td>VAMP7</td>\n",
       "      <td>ENSG00000124333</td>\n",
       "      <td>ENSG00000124333.13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19727</th>\n",
       "      <td>IL9R</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>ENSG00000124334.15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19728</th>\n",
       "      <td>SPRY3</td>\n",
       "      <td>ENSG00000168939</td>\n",
       "      <td>ENSG00000168939.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19045 rows Ã— 624 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_symbol       ensembl_id ensembl_id_with_version  TCGA-04-1517-01A  \\\n",
       "0          CDK11A  ENSG00000008128      ENSG00000008128.21                -1   \n",
       "1            NADK  ENSG00000008130      ENSG00000008130.14                -1   \n",
       "2           PRKCZ  ENSG00000067606      ENSG00000067606.14                -1   \n",
       "3            GNB1  ENSG00000078369      ENSG00000078369.16                -1   \n",
       "4            SDF4  ENSG00000078808      ENSG00000078808.15                -1   \n",
       "...           ...              ...                     ...               ...   \n",
       "19724       H2AB3  ENSG00000277745       ENSG00000277745.1                 0   \n",
       "19725       H2AB2  ENSG00000277858       ENSG00000277858.1                 0   \n",
       "19726       VAMP7  ENSG00000124333      ENSG00000124333.13                 0   \n",
       "19727        IL9R  ENSG00000124334      ENSG00000124334.15                 0   \n",
       "19728       SPRY3  ENSG00000168939       ENSG00000168939.9                 0   \n",
       "\n",
       "       TCGA-24-1556-01A  TCGA-36-1571-01A  TCGA-57-1582-01A  TCGA-25-1628-01A  \\\n",
       "0                     0                 0                 1                 0   \n",
       "1                     0                 0                 1                 0   \n",
       "2                     0                 0                 1                 0   \n",
       "3                     0                 0                 1                 0   \n",
       "4                     0                 0                 1                 0   \n",
       "...                 ...               ...               ...               ...   \n",
       "19724                 0                 0                -1                 0   \n",
       "19725                 0                 0                -1                 0   \n",
       "19726                 0                 0                -1                 0   \n",
       "19727                 0                 0                -1                 0   \n",
       "19728                 0                 0                -1                 0   \n",
       "\n",
       "       TCGA-09-1661-01B  TCGA-04-1519-01A  ...  TCGA-72-4240-01A.1  \\\n",
       "0                    -1                -1  ...                   0   \n",
       "1                    -1                -1  ...                   0   \n",
       "2                    -1                -1  ...                   0   \n",
       "3                    -1                -1  ...                   0   \n",
       "4                    -1                -1  ...                   0   \n",
       "...                 ...               ...  ...                 ...   \n",
       "19724                 0                 1  ...                   1   \n",
       "19725                 0                 1  ...                   1   \n",
       "19726                 0                 1  ...                   0   \n",
       "19727                 0                 1  ...                   0   \n",
       "19728                 0                 1  ...                   0   \n",
       "\n",
       "       TCGA-72-4232-01A.1  TCGA-72-4241-01A.1  TCGA-72-4233-01A.1  \\\n",
       "0                       0                   1                   1   \n",
       "1                       0                   1                   1   \n",
       "2                       0                   1                   1   \n",
       "3                       0                   1                   1   \n",
       "4                       0                   1                   1   \n",
       "...                   ...                 ...                 ...   \n",
       "19724                   0                   0                   1   \n",
       "19725                   0                   0                   1   \n",
       "19726                   0                   0                   1   \n",
       "19727                   0                   0                   1   \n",
       "19728                   0                   0                   1   \n",
       "\n",
       "       TCGA-72-4234-01A.1  TCGA-72-4235-01A.1  TCGA-72-4236-01A.1  \\\n",
       "0                       0                  -1                   1   \n",
       "1                       0                  -1                   1   \n",
       "2                       0                  -1                   1   \n",
       "3                       0                  -1                   1   \n",
       "4                       0                  -1                   1   \n",
       "...                   ...                 ...                 ...   \n",
       "19724                   1                   0                   0   \n",
       "19725                   1                   0                   0   \n",
       "19726                   1                   0                   0   \n",
       "19727                   1                   0                   0   \n",
       "19728                   1                   0                   0   \n",
       "\n",
       "       TCGA-72-4237-01A.1  TCGA-72-4238-01A.1  has_gene_symbol  \n",
       "0                       0                   1             True  \n",
       "1                       0                   1             True  \n",
       "2                       0                   1             True  \n",
       "3                       0                   1             True  \n",
       "4                       0                   1             True  \n",
       "...                   ...                 ...              ...  \n",
       "19724                   0                   0             True  \n",
       "19725                   0                   0             True  \n",
       "19726                   0                   0             True  \n",
       "19727                   0                   0             True  \n",
       "19728                   0                   0             True  \n",
       "\n",
       "[19045 rows x 624 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "copy_number_data = pd.read_csv(copy_number_file, sep=\"\\t\")\n",
    "\n",
    "# Need to parse gene symbol from ensembl id\n",
    "copy_number_data = copy_number_data.rename(columns={\"Gene Symbol\": \"ensembl_id_with_version\"})\n",
    "copy_number_data.insert(0,\n",
    "                        \"ensembl_id\",\n",
    "                        copy_number_data[\"ensembl_id_with_version\"].apply(\n",
    "                            lambda ensembl_id_with_version: ensembl_id_with_version.split(\".\")[0]\n",
    "                        ))\n",
    "copy_number_data[\"has_gene_symbol\"] = copy_number_data[\"ensembl_id\"].apply(\n",
    "        lambda ensembl_id: ensembl_id in ensembl_to_gene_symbol_df[\"ensembl_gene_id\"].values\n",
    "    )\n",
    "copy_number_data = copy_number_data[copy_number_data[\"has_gene_symbol\"]]\n",
    "copy_number_data.insert(0,\n",
    "                        \"gene_symbol\",\n",
    "                        copy_number_data[\"ensembl_id\"].apply(\n",
    "                            lambda ensembl_id: ensembl_to_gene_symbol_df[ensembl_to_gene_symbol_df[\"ensembl_gene_id\"] == ensembl_id][\"external_gene_name\"].values[0]\n",
    "                        ))\n",
    "\n",
    "display(copy_number_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-04-1517-01A</th>\n",
       "      <th>TCGA-24-1556-01A</th>\n",
       "      <th>TCGA-36-1571-01A</th>\n",
       "      <th>TCGA-57-1582-01A</th>\n",
       "      <th>TCGA-25-1628-01A</th>\n",
       "      <th>TCGA-09-1661-01B</th>\n",
       "      <th>TCGA-04-1519-01A</th>\n",
       "      <th>TCGA-24-1557-01A</th>\n",
       "      <th>TCGA-36-1574-01A</th>\n",
       "      <th>TCGA-57-1583-01A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-72-4240-01A.1</th>\n",
       "      <th>TCGA-72-4232-01A.1</th>\n",
       "      <th>TCGA-72-4241-01A.1</th>\n",
       "      <th>TCGA-72-4233-01A.1</th>\n",
       "      <th>TCGA-72-4234-01A.1</th>\n",
       "      <th>TCGA-72-4235-01A.1</th>\n",
       "      <th>TCGA-72-4236-01A.1</th>\n",
       "      <th>TCGA-72-4237-01A.1</th>\n",
       "      <th>TCGA-72-4238-01A.1</th>\n",
       "      <th>has_gene_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19724</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19725</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19726</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19727</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19728</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19045 rows Ã— 621 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TCGA-04-1517-01A  TCGA-24-1556-01A  TCGA-36-1571-01A  TCGA-57-1582-01A  \\\n",
       "0                    -1                 0                 0                 1   \n",
       "1                    -1                 0                 0                 1   \n",
       "2                    -1                 0                 0                 1   \n",
       "3                    -1                 0                 0                 1   \n",
       "4                    -1                 0                 0                 1   \n",
       "...                 ...               ...               ...               ...   \n",
       "19724                 0                 0                 0                -1   \n",
       "19725                 0                 0                 0                -1   \n",
       "19726                 0                 0                 0                -1   \n",
       "19727                 0                 0                 0                -1   \n",
       "19728                 0                 0                 0                -1   \n",
       "\n",
       "       TCGA-25-1628-01A  TCGA-09-1661-01B  TCGA-04-1519-01A  TCGA-24-1557-01A  \\\n",
       "0                     0                -1                -1                 1   \n",
       "1                     0                -1                -1                 1   \n",
       "2                     0                -1                -1                 1   \n",
       "3                     0                -1                -1                 1   \n",
       "4                     0                -1                -1                 1   \n",
       "...                 ...               ...               ...               ...   \n",
       "19724                 0                 0                 1                 0   \n",
       "19725                 0                 0                 1                 0   \n",
       "19726                 0                 0                 1                 0   \n",
       "19727                 0                 0                 1                 0   \n",
       "19728                 0                 0                 1                 0   \n",
       "\n",
       "       TCGA-36-1574-01A  TCGA-57-1583-01A  ...  TCGA-72-4240-01A.1  \\\n",
       "0                     0                 0  ...                   0   \n",
       "1                     0                 0  ...                   0   \n",
       "2                     0                 0  ...                   0   \n",
       "3                     0                 0  ...                   0   \n",
       "4                     0                 0  ...                   0   \n",
       "...                 ...               ...  ...                 ...   \n",
       "19724                 0                 0  ...                   1   \n",
       "19725                 0                 0  ...                   1   \n",
       "19726                 0                 0  ...                   0   \n",
       "19727                 0                 0  ...                   0   \n",
       "19728                 0                 0  ...                   0   \n",
       "\n",
       "       TCGA-72-4232-01A.1  TCGA-72-4241-01A.1  TCGA-72-4233-01A.1  \\\n",
       "0                       0                   1                   1   \n",
       "1                       0                   1                   1   \n",
       "2                       0                   1                   1   \n",
       "3                       0                   1                   1   \n",
       "4                       0                   1                   1   \n",
       "...                   ...                 ...                 ...   \n",
       "19724                   0                   0                   1   \n",
       "19725                   0                   0                   1   \n",
       "19726                   0                   0                   1   \n",
       "19727                   0                   0                   1   \n",
       "19728                   0                   0                   1   \n",
       "\n",
       "       TCGA-72-4234-01A.1  TCGA-72-4235-01A.1  TCGA-72-4236-01A.1  \\\n",
       "0                       0                  -1                   1   \n",
       "1                       0                  -1                   1   \n",
       "2                       0                  -1                   1   \n",
       "3                       0                  -1                   1   \n",
       "4                       0                  -1                   1   \n",
       "...                   ...                 ...                 ...   \n",
       "19724                   1                   0                   0   \n",
       "19725                   1                   0                   0   \n",
       "19726                   1                   0                   0   \n",
       "19727                   1                   0                   0   \n",
       "19728                   1                   0                   0   \n",
       "\n",
       "       TCGA-72-4237-01A.1  TCGA-72-4238-01A.1  has_gene_symbol  \n",
       "0                       0                   1             True  \n",
       "1                       0                   1             True  \n",
       "2                       0                   1             True  \n",
       "3                       0                   1             True  \n",
       "4                       0                   1             True  \n",
       "...                   ...                 ...              ...  \n",
       "19724                   0                   0             True  \n",
       "19725                   0                   0             True  \n",
       "19726                   0                   0             True  \n",
       "19727                   0                   0             True  \n",
       "19728                   0                   0             True  \n",
       "\n",
       "[19045 rows x 621 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "0     9007501\n",
       "1     1622832\n",
       "-1    1196612\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Examine CN data to try and interpret\n",
    "cn_only = copy_number_data.iloc[ : , 3 : ]\n",
    "display(cn_only)\n",
    "# Count the number of each type of copy number alteration\n",
    "stacked = cn_only.stack()\n",
    "\n",
    "# Count the occurrences of each value\n",
    "value_counts = stacked.value_counts()\n",
    "display(value_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Infer deletion of genes\n",
    "\n",
    "Infer deletion from log(counts+1). Original intention was to infer from copy number data, but it was uclear how to interpret the copy number data format.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num samples with TP53 deletions: 0/376\n",
      "Num samples with KRAS deletions: 0/376\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>TP53</th>\n",
       "      <th>tp53_deletion</th>\n",
       "      <th>kras_deletion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-04-1331</td>\n",
       "      <td>11.348175</td>\n",
       "      <td>10.115044</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-04-1332</td>\n",
       "      <td>10.540128</td>\n",
       "      <td>12.690871</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-04-1338</td>\n",
       "      <td>10.926296</td>\n",
       "      <td>12.447858</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-04-1341</td>\n",
       "      <td>11.886840</td>\n",
       "      <td>12.243174</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-04-1343</td>\n",
       "      <td>11.016808</td>\n",
       "      <td>11.253847</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>TCGA-61-2111</td>\n",
       "      <td>12.685406</td>\n",
       "      <td>12.276124</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>TCGA-61-2113</td>\n",
       "      <td>17.447196</td>\n",
       "      <td>13.516192</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>TCGA-OY-A56Q</td>\n",
       "      <td>12.386401</td>\n",
       "      <td>10.848623</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>TCGA-VG-A8LO</td>\n",
       "      <td>11.060696</td>\n",
       "      <td>13.300067</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>TCGA-WR-A838</td>\n",
       "      <td>13.273359</td>\n",
       "      <td>13.589300</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>376 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     submitter_id       KRAS       TP53  tp53_deletion  kras_deletion\n",
       "0    TCGA-04-1331  11.348175  10.115044          False          False\n",
       "1    TCGA-04-1332  10.540128  12.690871          False          False\n",
       "2    TCGA-04-1338  10.926296  12.447858          False          False\n",
       "3    TCGA-04-1341  11.886840  12.243174          False          False\n",
       "4    TCGA-04-1343  11.016808  11.253847          False          False\n",
       "..            ...        ...        ...            ...            ...\n",
       "371  TCGA-61-2111  12.685406  12.276124          False          False\n",
       "372  TCGA-61-2113  17.447196  13.516192          False          False\n",
       "373  TCGA-OY-A56Q  12.386401  10.848623          False          False\n",
       "374  TCGA-VG-A8LO  11.060696  13.300067          False          False\n",
       "375  TCGA-WR-A838  13.273359  13.589300          False          False\n",
       "\n",
       "[376 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check if any samples have 0 counts of TP53 or KRAS\n",
    "deletions_data = log2_counts.loc[log2_counts[\"gene_symbol\"].isin([\"TP53\", \"KRAS\"]), :]\n",
    "deletions_data.set_index(\"gene_symbol\", inplace=True)\n",
    "deletions_data = deletions_data.loc[:, deletions_data.columns.str.contains(\"TCGA\")]\n",
    "deletions_data = deletions_data.T.rename_axis(\"submitter_id\").reset_index().rename_axis(None, axis=1)\n",
    "deletions_data[\"tp53_deletion\"] = deletions_data[\"TP53\"].lt(5)\n",
    "deletions_data[\"kras_deletion\"] = deletions_data[\"KRAS\"].lt(5)\n",
    "\n",
    "print(f\"Num samples with TP53 deletions: {deletions_data['tp53_deletion'].sum()}/{deletions_data.shape[0]}\")\n",
    "print(f\"Num samples with KRAS deletions: {deletions_data['kras_deletion'].sum()}/{deletions_data.shape[0]}\")\n",
    "display(deletions_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspect metadata\n",
    "\n",
    "Only keep samples we have metadata, RNAseq, and mutations data for. (Don't need to have copy number data since we didn't end up using it to determine deletions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num samples with all data: 271/422\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primary_site</th>\n",
       "      <th>submitter_slide_ids</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>case_id</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>summary.file_count</th>\n",
       "      <th>summary.experimental_strategies</th>\n",
       "      <th>project.project_id</th>\n",
       "      <th>project.program.name</th>\n",
       "      <th>demographic.race</th>\n",
       "      <th>demographic.gender</th>\n",
       "      <th>demographic.ethnicity</th>\n",
       "      <th>demographic.vital_status</th>\n",
       "      <th>demographic.days_to_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-0920-01A-01-BS1, TCGA-13-0920-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>85a85a11-7200-4e96-97af-6ba26d680d59</td>\n",
       "      <td>TCGA-13-0920</td>\n",
       "      <td>[{'age_at_diagnosis': 24064}]</td>\n",
       "      <td>84</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1484.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-24-2033-01A-01-BS1, TCGA-24-2033-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>872d2922-7292-4681-adb7-d3b267eccbe7</td>\n",
       "      <td>TCGA-24-2033</td>\n",
       "      <td>[{'age_at_diagnosis': 31977}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>562.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-61-2110-11A-01-TS1, TCGA-61-2110-11A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>88180134-710f-46ea-9b06-5e5d860d6d9f</td>\n",
       "      <td>TCGA-61-2110</td>\n",
       "      <td>[{'age_at_diagnosis': 20718}]</td>\n",
       "      <td>69</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-29-1696-01A-01-TS1, TCGA-29-1696-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>8a98a6e6-b763-4824-858b-fd2738e6c9a3</td>\n",
       "      <td>TCGA-29-1696</td>\n",
       "      <td>[{'age_at_diagnosis': 15818}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1032.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-0911-01A-01-TS1, TCGA-13-0911-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>8cad4217-5699-4735-9be3-fc0015a8d262</td>\n",
       "      <td>TCGA-13-0911</td>\n",
       "      <td>[{'age_at_diagnosis': 20276}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-61-2102-01A-01-TS1, TCGA-61-2102-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>7a08efd0-f984-430e-a8eb-1881047214b6</td>\n",
       "      <td>TCGA-61-2102</td>\n",
       "      <td>[{'age_at_diagnosis': 27256}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>197.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-0804-01A-01-TS1, TCGA-13-0804-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>7e34d3c1-1fab-4326-9a69-4260d2bac558</td>\n",
       "      <td>TCGA-13-0804</td>\n",
       "      <td>[{'age_at_diagnosis': 27021}]</td>\n",
       "      <td>84</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1073.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-25-2042-01A-01-BS1, TCGA-25-2042-01Z-00-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>818dc159-aba4-46bc-a4ed-68ee0f8c4461</td>\n",
       "      <td>TCGA-25-2042</td>\n",
       "      <td>[{'age_at_diagnosis': 22127}]</td>\n",
       "      <td>69</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>american indian or alaska native</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Dead</td>\n",
       "      <td>396.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-13-1512-01A-01-TS1, TCGA-13-1512-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>82093ed9-a3c8-4e34-931f-4ec7ae745711</td>\n",
       "      <td>TCGA-13-1512</td>\n",
       "      <td>[{'age_at_diagnosis': 18134}]</td>\n",
       "      <td>70</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>Ovary</td>\n",
       "      <td>[TCGA-29-1768-01A-01-BS1, TCGA-29-1768-01A-01-...</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>838693d5-1ae0-4d75-834b-e1b89b96b0ed</td>\n",
       "      <td>TCGA-29-1768</td>\n",
       "      <td>[{'age_at_diagnosis': 18396}]</td>\n",
       "      <td>68</td>\n",
       "      <td>[WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...</td>\n",
       "      <td>TCGA-OV</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>952.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>271 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    primary_site                                submitter_slide_ids  \\\n",
       "0          Ovary  [TCGA-13-0920-01A-01-BS1, TCGA-13-0920-01A-01-...   \n",
       "2          Ovary  [TCGA-24-2033-01A-01-BS1, TCGA-24-2033-01A-01-...   \n",
       "3          Ovary  [TCGA-61-2110-11A-01-TS1, TCGA-61-2110-11A-01-...   \n",
       "4          Ovary  [TCGA-29-1696-01A-01-TS1, TCGA-29-1696-01A-01-...   \n",
       "6          Ovary  [TCGA-13-0911-01A-01-TS1, TCGA-13-0911-01A-01-...   \n",
       "..           ...                                                ...   \n",
       "416        Ovary  [TCGA-61-2102-01A-01-TS1, TCGA-61-2102-01A-01-...   \n",
       "418        Ovary  [TCGA-13-0804-01A-01-TS1, TCGA-13-0804-01A-01-...   \n",
       "419        Ovary  [TCGA-25-2042-01A-01-BS1, TCGA-25-2042-01Z-00-...   \n",
       "420        Ovary  [TCGA-13-1512-01A-01-TS1, TCGA-13-1512-01A-01-...   \n",
       "421        Ovary  [TCGA-29-1768-01A-01-BS1, TCGA-29-1768-01A-01-...   \n",
       "\n",
       "                              disease_type  \\\n",
       "0    Cystic, Mucinous and Serous Neoplasms   \n",
       "2    Cystic, Mucinous and Serous Neoplasms   \n",
       "3    Cystic, Mucinous and Serous Neoplasms   \n",
       "4    Cystic, Mucinous and Serous Neoplasms   \n",
       "6    Cystic, Mucinous and Serous Neoplasms   \n",
       "..                                     ...   \n",
       "416  Cystic, Mucinous and Serous Neoplasms   \n",
       "418  Cystic, Mucinous and Serous Neoplasms   \n",
       "419  Cystic, Mucinous and Serous Neoplasms   \n",
       "420  Cystic, Mucinous and Serous Neoplasms   \n",
       "421  Cystic, Mucinous and Serous Neoplasms   \n",
       "\n",
       "                                  case_id  submitter_id  \\\n",
       "0    85a85a11-7200-4e96-97af-6ba26d680d59  TCGA-13-0920   \n",
       "2    872d2922-7292-4681-adb7-d3b267eccbe7  TCGA-24-2033   \n",
       "3    88180134-710f-46ea-9b06-5e5d860d6d9f  TCGA-61-2110   \n",
       "4    8a98a6e6-b763-4824-858b-fd2738e6c9a3  TCGA-29-1696   \n",
       "6    8cad4217-5699-4735-9be3-fc0015a8d262  TCGA-13-0911   \n",
       "..                                    ...           ...   \n",
       "416  7a08efd0-f984-430e-a8eb-1881047214b6  TCGA-61-2102   \n",
       "418  7e34d3c1-1fab-4326-9a69-4260d2bac558  TCGA-13-0804   \n",
       "419  818dc159-aba4-46bc-a4ed-68ee0f8c4461  TCGA-25-2042   \n",
       "420  82093ed9-a3c8-4e34-931f-4ec7ae745711  TCGA-13-1512   \n",
       "421  838693d5-1ae0-4d75-834b-e1b89b96b0ed  TCGA-29-1768   \n",
       "\n",
       "                         diagnoses  summary.file_count  \\\n",
       "0    [{'age_at_diagnosis': 24064}]                  84   \n",
       "2    [{'age_at_diagnosis': 31977}]                  70   \n",
       "3    [{'age_at_diagnosis': 20718}]                  69   \n",
       "4    [{'age_at_diagnosis': 15818}]                  70   \n",
       "6    [{'age_at_diagnosis': 20276}]                  70   \n",
       "..                             ...                 ...   \n",
       "416  [{'age_at_diagnosis': 27256}]                  70   \n",
       "418  [{'age_at_diagnosis': 27021}]                  84   \n",
       "419  [{'age_at_diagnosis': 22127}]                  69   \n",
       "420  [{'age_at_diagnosis': 18134}]                  70   \n",
       "421  [{'age_at_diagnosis': 18396}]                  68   \n",
       "\n",
       "                       summary.experimental_strategies project.project_id  \\\n",
       "0    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "2    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "3    [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "4    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "6    [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "..                                                 ...                ...   \n",
       "416  [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "418  [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "419  [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "420  [WXS, miRNA-Seq, RNA-Seq, WGS, Genotyping Arra...            TCGA-OV   \n",
       "421  [WXS, miRNA-Seq, RNA-Seq, Genotyping Array, Me...            TCGA-OV   \n",
       "\n",
       "    project.program.name                  demographic.race demographic.gender  \\\n",
       "0                   TCGA                             white             female   \n",
       "2                   TCGA                             white             female   \n",
       "3                   TCGA                             white             female   \n",
       "4                   TCGA                             white             female   \n",
       "6                   TCGA                             white             female   \n",
       "..                   ...                               ...                ...   \n",
       "416                 TCGA                             white             female   \n",
       "418                 TCGA                             white             female   \n",
       "419                 TCGA  american indian or alaska native             female   \n",
       "420                 TCGA                             white             female   \n",
       "421                 TCGA                             white             female   \n",
       "\n",
       "      demographic.ethnicity demographic.vital_status  \\\n",
       "0    not hispanic or latino                     Dead   \n",
       "2              not reported                     Dead   \n",
       "3    not hispanic or latino                     Dead   \n",
       "4              not reported                     Dead   \n",
       "6    not hispanic or latino                     Dead   \n",
       "..                      ...                      ...   \n",
       "416            not reported                     Dead   \n",
       "418  not hispanic or latino                     Dead   \n",
       "419  not hispanic or latino                     Dead   \n",
       "420  not hispanic or latino                    Alive   \n",
       "421            not reported                     Dead   \n",
       "\n",
       "     demographic.days_to_death  \n",
       "0                       1484.0  \n",
       "2                        562.0  \n",
       "3                       1354.0  \n",
       "4                       1032.0  \n",
       "6                       1355.0  \n",
       "..                         ...  \n",
       "416                      197.0  \n",
       "418                     1073.0  \n",
       "419                      396.0  \n",
       "420                        NaN  \n",
       "421                      952.0  \n",
       "\n",
       "[271 rows x 15 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "relevant_samples = set(metadata[\"submitter_id\"]) & set(log2_counts.columns) & set(mutations[\"submitter_id\"]) & set(deletions_data[\"submitter_id\"])\n",
    "relevant_metadata = metadata.loc[metadata['submitter_id'].isin(relevant_samples)]\n",
    "print(f\"Num samples with all data: {relevant_metadata.shape[0]}/{metadata.shape[0]}\")\n",
    "\n",
    "# Check that all experimental strategies include RNA-Seq\n",
    "assert all([\"RNA-Seq\" in exp for exp in relevant_metadata[\"summary.experimental_strategies\"]])\n",
    "\n",
    "# Check that all are Cystic, Mucinous and Serous Neoplasms\n",
    "assert all([\"Cystic, Mucinous and Serous Neoplasms\" in disease_type for disease_type in relevant_metadata[\"disease_type\"]])\n",
    "\n",
    "display(relevant_metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset data\n",
    "\n",
    "Only keep samples that have TP53 mutations and do not have KRAS mutations in an effort to make sure that we have only HGSC celllines (KRAS indicates LGSC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assert that there are no duplicate submitter ids in mutations\n",
    "assert len(set(mutations[\"submitter_id\"])) == len(mutations[\"submitter_id\"])\n",
    "# Assert that there are no duplicate submitter ids in log2_counts\n",
    "assert len(set(log2_counts.columns)) == len(log2_counts.columns)\n",
    "# Assert that there are no duplicate submitter ids in deletions_data\n",
    "assert len(set(deletions_data[\"submitter_id\"])) == len(deletions_data[\"submitter_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num samples with TP53 mutations/deletions and no KRAS mutations/deletions: 191/271\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>tp53_mutant</th>\n",
       "      <th>tp53_deletion</th>\n",
       "      <th>kras_mutant</th>\n",
       "      <th>kras_deletion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-13-0920</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-24-2033</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-61-2110</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-61-2111</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TCGA-24-2254</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>TCGA-59-2352</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>TCGA-61-2102</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>TCGA-13-0804</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>TCGA-13-1512</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>TCGA-29-1768</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     submitter_id  tp53_mutant  tp53_deletion  kras_mutant  kras_deletion\n",
       "0    TCGA-13-0920         True          False        False          False\n",
       "1    TCGA-24-2033         True          False        False          False\n",
       "2    TCGA-61-2110         True          False        False          False\n",
       "6    TCGA-61-2111         True          False        False          False\n",
       "10   TCGA-24-2254         True          False        False          False\n",
       "..            ...          ...            ...          ...            ...\n",
       "262  TCGA-59-2352         True          False        False          False\n",
       "266  TCGA-61-2102         True          False        False          False\n",
       "267  TCGA-13-0804         True          False        False          False\n",
       "269  TCGA-13-1512         True          False        False          False\n",
       "270  TCGA-29-1768         True          False        False          False\n",
       "\n",
       "[191 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "relevant_samples = relevant_metadata[['submitter_id']].merge(\n",
    "    mutations[['submitter_id', 'tp53_mutant', 'kras_mutant']], on='submitter_id', how='left'\n",
    ").merge(\n",
    "    deletions_data[['submitter_id', 'tp53_deletion', 'kras_deletion']], on='submitter_id', how='left'\n",
    ")\n",
    "relevant_samples = relevant_samples[['submitter_id', 'tp53_mutant', 'tp53_deletion', 'kras_mutant', 'kras_deletion']]\n",
    "\n",
    "subset_samples = relevant_samples[\n",
    "    # Has a TP53 deletion/mutation (indicator of HGSC)\n",
    "    (relevant_samples[\"tp53_deletion\"] | relevant_samples[\"tp53_mutant\"]) &\n",
    "    # And does not have a kras deletion/mutation (indicator of LGSC)\n",
    "    ~ (relevant_samples[\"kras_deletion\"] | relevant_samples[\"kras_mutant\"])\n",
    "]\n",
    "\n",
    "print(f\"Num samples with TP53 mutations/deletions and no KRAS mutations/deletions: {subset_samples.shape[0]}/{relevant_samples.shape[0]}\")\n",
    "display(subset_samples)\n",
    "\n",
    "# Save subset samples to a file\n",
    "subset_samples.to_csv(f\"{tcga_data_folder}/confident_hgsc_samples.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset RNAseq data based for high confidence HGSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-13-0920</th>\n",
       "      <th>TCGA-24-2033</th>\n",
       "      <th>TCGA-61-2110</th>\n",
       "      <th>TCGA-61-2111</th>\n",
       "      <th>TCGA-24-2254</th>\n",
       "      <th>TCGA-24-1428</th>\n",
       "      <th>TCGA-24-2280</th>\n",
       "      <th>TCGA-24-1843</th>\n",
       "      <th>TCGA-23-1110</th>\n",
       "      <th>TCGA-10-0937</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-24-1552</th>\n",
       "      <th>TCGA-25-1321</th>\n",
       "      <th>TCGA-13-0888</th>\n",
       "      <th>TCGA-04-1331</th>\n",
       "      <th>TCGA-13-1511</th>\n",
       "      <th>TCGA-59-2352</th>\n",
       "      <th>TCGA-61-2102</th>\n",
       "      <th>TCGA-13-0804</th>\n",
       "      <th>TCGA-13-1512</th>\n",
       "      <th>TCGA-29-1768</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>11.829326</td>\n",
       "      <td>12.405673</td>\n",
       "      <td>9.952741</td>\n",
       "      <td>11.976922</td>\n",
       "      <td>12.487338</td>\n",
       "      <td>13.346929</td>\n",
       "      <td>11.777255</td>\n",
       "      <td>12.191985</td>\n",
       "      <td>10.302639</td>\n",
       "      <td>11.919608</td>\n",
       "      <td>...</td>\n",
       "      <td>11.438792</td>\n",
       "      <td>12.175238</td>\n",
       "      <td>10.766529</td>\n",
       "      <td>12.804131</td>\n",
       "      <td>12.615170</td>\n",
       "      <td>11.976206</td>\n",
       "      <td>11.413099</td>\n",
       "      <td>11.454299</td>\n",
       "      <td>12.386132</td>\n",
       "      <td>11.564149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.087463</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>4.459432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.459432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.523562</td>\n",
       "      <td>5.426265</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.906891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>10.736402</td>\n",
       "      <td>11.681238</td>\n",
       "      <td>10.800091</td>\n",
       "      <td>11.249706</td>\n",
       "      <td>10.576484</td>\n",
       "      <td>12.022021</td>\n",
       "      <td>11.489346</td>\n",
       "      <td>12.463013</td>\n",
       "      <td>10.991522</td>\n",
       "      <td>10.172428</td>\n",
       "      <td>...</td>\n",
       "      <td>11.494856</td>\n",
       "      <td>11.937006</td>\n",
       "      <td>9.975848</td>\n",
       "      <td>11.669771</td>\n",
       "      <td>10.973697</td>\n",
       "      <td>10.830515</td>\n",
       "      <td>10.632995</td>\n",
       "      <td>9.652845</td>\n",
       "      <td>11.556985</td>\n",
       "      <td>10.731319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>9.116344</td>\n",
       "      <td>9.339850</td>\n",
       "      <td>9.398744</td>\n",
       "      <td>10.135709</td>\n",
       "      <td>9.415742</td>\n",
       "      <td>10.687376</td>\n",
       "      <td>9.856426</td>\n",
       "      <td>10.115044</td>\n",
       "      <td>7.994353</td>\n",
       "      <td>7.592457</td>\n",
       "      <td>...</td>\n",
       "      <td>9.541097</td>\n",
       "      <td>10.802516</td>\n",
       "      <td>8.144658</td>\n",
       "      <td>10.364135</td>\n",
       "      <td>10.344296</td>\n",
       "      <td>9.348728</td>\n",
       "      <td>9.957102</td>\n",
       "      <td>8.098032</td>\n",
       "      <td>10.144658</td>\n",
       "      <td>8.430453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRRM</th>\n",
       "      <td>8.647458</td>\n",
       "      <td>9.497852</td>\n",
       "      <td>8.876517</td>\n",
       "      <td>9.245553</td>\n",
       "      <td>8.909893</td>\n",
       "      <td>10.395534</td>\n",
       "      <td>9.189825</td>\n",
       "      <td>10.182394</td>\n",
       "      <td>7.894818</td>\n",
       "      <td>6.442943</td>\n",
       "      <td>...</td>\n",
       "      <td>9.245553</td>\n",
       "      <td>9.930737</td>\n",
       "      <td>7.055282</td>\n",
       "      <td>9.812177</td>\n",
       "      <td>8.867279</td>\n",
       "      <td>8.523562</td>\n",
       "      <td>10.203348</td>\n",
       "      <td>7.247928</td>\n",
       "      <td>9.483816</td>\n",
       "      <td>7.546894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>None</th>\n",
       "      <td>21.648527</td>\n",
       "      <td>22.301199</td>\n",
       "      <td>22.747116</td>\n",
       "      <td>22.056261</td>\n",
       "      <td>22.922068</td>\n",
       "      <td>23.909925</td>\n",
       "      <td>23.488641</td>\n",
       "      <td>22.820357</td>\n",
       "      <td>21.135925</td>\n",
       "      <td>20.713963</td>\n",
       "      <td>...</td>\n",
       "      <td>21.657224</td>\n",
       "      <td>23.118446</td>\n",
       "      <td>22.354941</td>\n",
       "      <td>23.565784</td>\n",
       "      <td>23.082234</td>\n",
       "      <td>22.039606</td>\n",
       "      <td>21.584929</td>\n",
       "      <td>21.979689</td>\n",
       "      <td>21.552433</td>\n",
       "      <td>20.889634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>None</th>\n",
       "      <td>20.723864</td>\n",
       "      <td>21.165166</td>\n",
       "      <td>21.386072</td>\n",
       "      <td>20.646611</td>\n",
       "      <td>20.809075</td>\n",
       "      <td>22.842148</td>\n",
       "      <td>21.363436</td>\n",
       "      <td>22.090146</td>\n",
       "      <td>19.815744</td>\n",
       "      <td>20.599656</td>\n",
       "      <td>...</td>\n",
       "      <td>21.313734</td>\n",
       "      <td>21.682592</td>\n",
       "      <td>20.399962</td>\n",
       "      <td>21.348947</td>\n",
       "      <td>21.768537</td>\n",
       "      <td>21.353379</td>\n",
       "      <td>20.672535</td>\n",
       "      <td>20.915839</td>\n",
       "      <td>21.583226</td>\n",
       "      <td>21.185491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>None</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>None</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>None</th>\n",
       "      <td>25.318455</td>\n",
       "      <td>25.082005</td>\n",
       "      <td>25.584612</td>\n",
       "      <td>24.721992</td>\n",
       "      <td>24.569189</td>\n",
       "      <td>26.552636</td>\n",
       "      <td>25.083444</td>\n",
       "      <td>26.208837</td>\n",
       "      <td>23.563361</td>\n",
       "      <td>25.039234</td>\n",
       "      <td>...</td>\n",
       "      <td>25.341984</td>\n",
       "      <td>25.962545</td>\n",
       "      <td>25.758646</td>\n",
       "      <td>25.853210</td>\n",
       "      <td>25.751119</td>\n",
       "      <td>24.936333</td>\n",
       "      <td>24.452542</td>\n",
       "      <td>24.341527</td>\n",
       "      <td>25.816958</td>\n",
       "      <td>26.486673</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60488 rows Ã— 191 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TCGA-13-0920  TCGA-24-2033  TCGA-61-2110  TCGA-61-2111  \\\n",
       "gene_symbol                                                           \n",
       "TSPAN6          11.829326     12.405673      9.952741     11.976922   \n",
       "TNMD             3.000000      2.321928      2.584963      0.000000   \n",
       "DPM1            10.736402     11.681238     10.800091     11.249706   \n",
       "SCYL3            9.116344      9.339850      9.398744     10.135709   \n",
       "FIRRM            8.647458      9.497852      8.876517      9.245553   \n",
       "...                   ...           ...           ...           ...   \n",
       "None            21.648527     22.301199     22.747116     22.056261   \n",
       "None            20.723864     21.165166     21.386072     20.646611   \n",
       "None             0.000000      0.000000      0.000000      0.000000   \n",
       "None             0.000000      0.000000      0.000000      0.000000   \n",
       "None            25.318455     25.082005     25.584612     24.721992   \n",
       "\n",
       "             TCGA-24-2254  TCGA-24-1428  TCGA-24-2280  TCGA-24-1843  \\\n",
       "gene_symbol                                                           \n",
       "TSPAN6          12.487338     13.346929     11.777255     12.191985   \n",
       "TNMD             4.087463      3.000000      4.459432      1.000000   \n",
       "DPM1            10.576484     12.022021     11.489346     12.463013   \n",
       "SCYL3            9.415742     10.687376      9.856426     10.115044   \n",
       "FIRRM            8.909893     10.395534      9.189825     10.182394   \n",
       "...                   ...           ...           ...           ...   \n",
       "None            22.922068     23.909925     23.488641     22.820357   \n",
       "None            20.809075     22.842148     21.363436     22.090146   \n",
       "None             0.000000      0.000000      0.000000      0.000000   \n",
       "None             0.000000      0.000000      0.000000      0.000000   \n",
       "None            24.569189     26.552636     25.083444     26.208837   \n",
       "\n",
       "             TCGA-23-1110  TCGA-10-0937  ...  TCGA-24-1552  TCGA-25-1321  \\\n",
       "gene_symbol                              ...                               \n",
       "TSPAN6          10.302639     11.919608  ...     11.438792     12.175238   \n",
       "TNMD             2.000000      2.000000  ...      1.000000      1.584963   \n",
       "DPM1            10.991522     10.172428  ...     11.494856     11.937006   \n",
       "SCYL3            7.994353      7.592457  ...      9.541097     10.802516   \n",
       "FIRRM            7.894818      6.442943  ...      9.245553      9.930737   \n",
       "...                   ...           ...  ...           ...           ...   \n",
       "None            21.135925     20.713963  ...     21.657224     23.118446   \n",
       "None            19.815744     20.599656  ...     21.313734     21.682592   \n",
       "None             0.000000      0.000000  ...      0.000000      0.000000   \n",
       "None             0.000000      0.000000  ...      0.000000      0.000000   \n",
       "None            23.563361     25.039234  ...     25.341984     25.962545   \n",
       "\n",
       "             TCGA-13-0888  TCGA-04-1331  TCGA-13-1511  TCGA-59-2352  \\\n",
       "gene_symbol                                                           \n",
       "TSPAN6          10.766529     12.804131     12.615170     11.976206   \n",
       "TNMD             1.000000      3.459432      1.000000      5.523562   \n",
       "DPM1             9.975848     11.669771     10.973697     10.830515   \n",
       "SCYL3            8.144658     10.364135     10.344296      9.348728   \n",
       "FIRRM            7.055282      9.812177      8.867279      8.523562   \n",
       "...                   ...           ...           ...           ...   \n",
       "None            22.354941     23.565784     23.082234     22.039606   \n",
       "None            20.399962     21.348947     21.768537     21.353379   \n",
       "None             0.000000      0.000000      0.000000      0.000000   \n",
       "None             0.000000      0.000000      0.000000      0.000000   \n",
       "None            25.758646     25.853210     25.751119     24.936333   \n",
       "\n",
       "             TCGA-61-2102  TCGA-13-0804  TCGA-13-1512  TCGA-29-1768  \n",
       "gene_symbol                                                          \n",
       "TSPAN6          11.413099     11.454299     12.386132     11.564149  \n",
       "TNMD             5.426265      1.584963      0.000000      4.906891  \n",
       "DPM1            10.632995      9.652845     11.556985     10.731319  \n",
       "SCYL3            9.957102      8.098032     10.144658      8.430453  \n",
       "FIRRM           10.203348      7.247928      9.483816      7.546894  \n",
       "...                   ...           ...           ...           ...  \n",
       "None            21.584929     21.979689     21.552433     20.889634  \n",
       "None            20.672535     20.915839     21.583226     21.185491  \n",
       "None             0.000000      0.000000      0.000000      0.000000  \n",
       "None             0.000000      0.000000      0.000000      0.000000  \n",
       "None            24.452542     24.341527     25.816958     26.486673  \n",
       "\n",
       "[60488 rows x 191 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "log2_counts_submitter_id = log2_counts.set_index(\"gene_symbol\").rename_axis(\"gene_symbol\")\n",
    "subset_log2counts = log2_counts_submitter_id[subset_samples[\"submitter_id\"]]\n",
    "display(subset_log2counts)\n",
    "\n",
    "subset_log2counts.to_csv(f\"{tcga_data_folder}/confident_hgsc_log2counts.tsv\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
